yarn.config.cjs

```diff
@@ -0,0 +1,654 @@
+ *
+ * @typedef {import('@yarnpkg/types').Yarn} Yarn
+ * @typedef {import('@yarnpkg/types').Yarn.Constraints.Workspace} Workspace
+ * @typedef {import('@yarnpkg/types').Yarn.Constraints.Dependency} Dependency
+ * @typedef {import('@yarnpkg/types').Yarn.Constraints.DependencyType} DependencyType
+ */
+
+module.exports = defineConfig({
+  async constraints({ Yarn }) {
+    const rootWorkspace = Yarn.workspace({ cwd: '.' });
+    if (rootWorkspace === null) {
+      throw new Error('Could not find root workspace');
+    }
+
+    const repositoryUri = rootWorkspace.manifest.repository.url.replace(
+      /\.git$/u,
+      '',
+    );
+
+    for (const workspace of Yarn.workspaces()) {
+      const workspaceBasename = getWorkspaceBasename(workspace);
+      const isChildWorkspace = workspace.cwd !== '.';
+      const isPrivate =
+        'private' in workspace.manifest && workspace.manifest.private === true;
+      const dependenciesByIdentAndType = getDependenciesByIdentAndType(
+        Yarn.dependencies({ workspace }),
+      );
+
+      // All packages must have a name.
+      expectWorkspaceField(workspace, 'name');
+
+      if (isChildWorkspace) {
+        // All non-root packages must have a name that matches its directory
+        // (e.g., a package in a workspace directory called `foo` must be called
+        // `@metamask/foo`).
+        expectWorkspaceField(
+          workspace,
+          'name',
+          `@metamask/${workspaceBasename}`,
+        );
+
+        // All non-root packages must have a version.
+        expectWorkspaceField(workspace, 'version');
+
+        // All non-root packages must have a description that ends in a period.
+        expectWorkspaceDescription(workspace);
+
+        // All non-root packages must have the same set of NPM keywords.
+        expectWorkspaceField(workspace, 'keywords', ['MetaMask', 'Ethereum']);
+
+        // All non-root packages must have a homepage URL that includes its name.
+        expectWorkspaceField(
+          workspace,
+          'homepage',
+          `${repositoryUri}/tree/main/packages/${workspaceBasename}#readme`,
+        );
+
+        // All non-root packages must have a URL for reporting bugs that points
+        // to the Issues page for the repository.
+        expectWorkspaceField(workspace, 'bugs.url', `${repositoryUri}/issues`);
+
+        // All non-root packages must specify a Git repository within the
+        // MetaMask GitHub organization.
+        expectWorkspaceField(workspace, 'repository.type', 'git');
+        expectWorkspaceField(
+          workspace,
+          'repository.url',
+          `${repositoryUri}.git`,
+        );
+
+        // If not specified, the license for all non-root packages must be MIT.
+        if (
+          workspace.manifest.license === null ||
+          workspace.manifest.license === undefined
+        ) {
+          expectWorkspaceField(workspace, 'license', 'MIT');
+        }
```

Comment:
the following rule is missing for the packages that are under ISC license
```
% The license for all published packages must be MIT unless otherwise specified.
gen_enforced_field(WorkspaceCwd, 'license', 'MIT') :-
  \+ workspace_field(WorkspaceCwd, 'private', true),
  WorkspaceCwd \= 'packages/json-rpc-engine',
  WorkspaceCwd \= 'packages/json-rpc-middleware-stream',
  WorkspaceCwd \= 'packages/permission-log-controller',
  WorkspaceCwd \= 'packages/eth-json-rpc-provider'.
% The following published packages use an ISC license instead of MIT.
gen_enforced_field(WorkspaceCwd, 'license', 'ISC') :-
  \+ workspace_field(WorkspaceCwd, 'private', true),
  (
    WorkspaceCwd == 'packages/json-rpc-engine' ;
    WorkspaceCwd == 'packages/json-rpc-middleware-stream' ;
    WorkspaceCwd == 'packages/eth-json-rpc-provider'
  ).
```

Reply:
I decided to update these checks slightly so that we don't have to keep track of which packages _aren't_ using MIT. My thought here was that if someone is adding a new package and forgets to set the license, the constraints will ask them to use MIT. However, if a package is already using a license, then that should be okay, and it can use whatever license it wants to. Does that make sense or do you think it is still worth it to track those exceptions here?

Reply:
I see your point. My only concern is that this could allow someone to unintentionally commit a package under a license we do not want to permit.

Reply:
I've updated this check to enforce MIT if `license` is missing, but to also make an exception for `json-rpc-engine`, `json-rpc-middleware-stream`, `permission-log-controller`, and `eth-json-rpc-provider`. I also realized that I could make sure that the LICENSE file was present, so I added a check for that and extracted the whole check to a function to clean up the code a bit more: [c51cc83](https://github.com/MetaMask/core/pull/4546/commits/c51cc83774c3dc510c022f5db7d3373d5c33bf6a)